core/src/smartcontracts/wasm.rs

```diff
@@ -69,25 +69,80 @@ pub mod import {
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
     /// Engine or linker could not be created
-    #[error("Runtime initialization failure: {0}")]
+    #[error("Runtime initialization failure")]
     Initialization(#[source] eyre::Report),
     /// Module could not be compiled or instantiated
-    #[error("Module instantiation failure: {0}")]
+    #[error("Module instantiation failure")]
     Instantiation(#[source] eyre::Report),
     /// Expected named export not found in module
-    #[error("Named export not found: {0}")]
+    #[error("Named export not found")]
     ExportNotFound(#[source] eyre::Report),
+    /// Error exporting memory from wasm to iroha
+    #[error("Failed to export memory")]
+    MemoryExport(#[from] MemoryExportError),
```

Comment:
this should be under `Initialization` or `Instantiation`? Also, this error is hardly observable because it would mean someone has tapped into our framework

Reply:
Currently we're trying to get this export i.e. on every call to `execute_query` or `execute_instruction`, so it's not an `Initialization` or `Instantiation` error.

But I'll try to refactor this

Reply:
ok, yeah. But can we get this error after initialization?

Reply:
I will try to call `get_memory()` once and store it in state, so that we can't get this error at real runtime

Reply:
Okay, actually it's not so trivial, because to get memory we should already have `State` instantiated, so it's not possible to store `wasmtime::Memory` without using `Option` which I don't like here.

And still we could get this error after initialization (it's wasm engine creation time) and after instantiation (it's concrete wasm blobl instantiation time). So I'd better keep it like it is.

> Also, this error is hardly observable because it would mean someone has tapped into our framework

Yes, but we still should handle this type of error, because user can submit any wierd wasm code. That's why I call it `InternalError` but I do not use panic, because it's not *Iroha* bug, it's admin mistake.

Reply:
Why are you sure that on wasm intialization step (which one, actually?) wasmtime also checks for `memory` export?

Reply:
I think if it would be like so then there would be a way to just call `memory()` on something

Reply:
I think you are wrong. The access to memory is name based, it's not engraved into the type system. The compiler cannot know if any particular access is correct. But we have established this invariant when the module was instantiated so we don't need to worry about it anymore. 

Also, if this error is possible at runtime, it is very appropriate to just `Trap`. This error cannot be handled in the program in any reasonable way

Reply:
I think you are wrong. The access to memory is name based, it's not engraved into the type system. The compiler cannot know if any particular access is correct. But we have established this invariant when the module was instantiated so we don't need to worry about it anymore. 

Also, if this error is possible at runtime, it is very appropriate to just `Trap`. This error cannot be handled in the program in any reasonable way

Reply:
>  But we have established this invariant when the module was instantiated so we don't need to worry about it anymore.

I don't see that in the code. We don't check that `memory` is a really valid memory at instantiation time.
Do you suggest to implement it and then just panic if we can't get memory later?

Reply:
I would recommend we revisit this after #1615 is implemented. If the code is dead, we should see it in full-coverage failure point test. 

I agree with @mversic about the general sentiment, I just don't think that reasoning about code with regards to memory safety is sufficient (If it were, Rust wouldn't exist). 

So I would recommend splitting this off into a separate ticket and closing the conversation.

Reply:
This error type was refactored, so check out new version and leave a comment for it. Closing this